hail/src/main/scala/is/hail/rvd/RVD.scala

```diff
@@ -114,14 +114,49 @@ class RVD(
   ): RVD = {
     require(newKey.forall(rowType.hasField))
     val nPreservedFields = typ.key.zip(newKey).takeWhile { case (l, r) => l == r }.length
-    require(!isSorted || nPreservedFields > 0 || newKey.isEmpty)
+    val maybeKeys = partitioner.keysIfOneToOne()
+    if (!(!isSorted || nPreservedFields > 0 || newKey.isEmpty || maybeKeys.isDefined)) {
+      throw new IllegalArgumentException(s"$isSorted, $nPreservedFields, $newKey, ${maybeKeys.isDefined}, ${typ}, ${partitioner}")
+    }
 
     if (nPreservedFields == newKey.length)
       this
     else if (isSorted)
-      truncateKey(newKey.take(nPreservedFields))
-        .extendKeyPreservesPartitioning(execCtx, newKey)
-        .checkKeyOrdering()
+      maybeKeys match {
+        case None =>
+          truncateKey(newKey.take(nPreservedFields))
+            .extendKeyPreservesPartitioning(execCtx, newKey)
+            .checkKeyOrdering()
+        case Some(keys) =>
+          val oldRVDType = typ
+          val oldKeyPType = oldRVDType.kType
+          val oldKeyVType = oldKeyPType.virtualType
+          val newRVDType = oldRVDType.copy(key = newKey)
+          val newKeyPType = newRVDType.kType
+          val newKeyVType = newKeyPType.virtualType
+          var keyInfo = keys.zipWithIndex.map { case (oldKey, partitionIndex) =>
+            val newKey = new SelectFieldsRow(oldKey, oldKeyPType, newKeyPType)
```

Comment:
Good points. They sent me off thinking a lot about RVDPartitioners. In the end I think the code as written is the semantics I need for dndarray.

On the first point, my intention is to have exactly this behavior. 

On the second point, I'll ensure we never use this no-shuffle, fast-path when we're extending the keys. I do not need to extend keys to support dndarray. Given that, I am to keep this change as simple as possible. I don't think coarsen is the right thing because I'm permuting the keys, `nPreservedFields` might be 0.

Reply:
> On the first point, my intention is to have exactly this behavior.

Right, but this is on the generic `key_by` path, and this is no longer an obvious optimization in all cases. I think my real question is: what is the new semantics for `key_by`? If I want to change my key from `[A, B]` to `[B]`, then probably it will shuffle and choose balanced partitions, keeping roughly the same amount of parallelism, but if the existing partitioner satisfies a somewhat obscure condition that I don't have much control over, it will instead coalesce partitions.

What if we gate this behavior behind a flag on `TableKeyBy`, and expose a way to opt in to the optimization in python?